### ISSUE TITLE
Return non-null, probably stack-emplaceable, value types

### ISSUE THREAD

**Issue by adetaylor** [google/autocxx#845]
Currently, a C++ function:

```cpp
Thing return_thing();
```

returns a `cxx::UniquePtr<Thing>` in Rust (unless `Thing` is declared to be a POD type, which we'll disregard).

This has three problems:
* It requires heap allocation. With #379 we increasingly allow non-trivial C++ objects to be kept on the Rust stack.
* On the C++ side we use `std::move`, so all types have to be movable.
* The `UniquePtr` may be null, so any access to it requires a dance like `my_thing.as_ref().unwrap()` even though we know _we_ wrapped it in a `UniquePtr` so we can guarantee it's not null.

Options include:
* An `autocxx::NotNullUniquePtr` newtype wrapper.
* Upstreaming something similar to `cxx`.
* Doing the full `moveit` monty and returning an `impl New<Output=Thing>`. This would allow stack, `UniquePtr` or `Box` emplacement of the value.

The third option is most flexible. It's fairly complex:
* We'd need to pass the address of the `MaybeUninit` across to C++ as an extra output parameter into each function call.
* We'd then need to call the C++ function as `*out_param = underlying_cpp_function()` which should always RVO.

Concerns:
* This is complex.
* Ergonomics. Right now, we can call:
```rust
takes_t(makes_t());
```

How ergonomic can we make that?

```rust
takes_t(moveit! { makes_t() })
```

```rust
takes_t(Box::emplace(makes_t())
```

```rust
takes_t(makes_t().emplace_box())
```

**Comment by bsilver8192** [comment]
I would prefer the C++ as `new (out_param) (underlying_cpp_function())`. Don't need to rely on RVO, and it avoids requiring the C++ type to have a move assignment operator. Having a move constructor but no move assignment operator is weird, but I don't think any part of autocxx cares right now, doing it with placement new reduces the scope of unusual C++ types that autocxx needs to worry about.

Could `takes_t(makes_t())` continue to work by genericizing `makes_t`? Something like `fn makes_t(t: impl Into<T&>) { cxxbridge::makes_t(t.into()); }`, and then allow calling that with any of the container types. Not sure if using `Into` (or maybe `AsRef`?) is a good idea, or if it creating a separate version of the same idea works out better.

**Comment by adetaylor** [comment]
`takes_t(makes_t())` would currently be `takes_t(impl ValueParam<T>)`. Unfortunately it turns out to be fiddly to `impl ValueParam<T> for T where T: New<Output=T>` which is what I really want to do, because Rust is concerned downstream crates might impl `New` for things that we have other `ValueParam` impls for already (e.g. `UniquePtr`).

It's possible that in the future `ValueParam` could just be replaced with `New`, so long as it turns out we can `impl New for UniquePtr<T> where T: MoveNew` or similar. TBD.

**Comment by adetaylor** [comment]
Fixed by #998.
